### PR TITLE
Changes that allow Mono build on Linux

### PR DESCRIPTION
**Pull request by ValMenn**
Most changes have to do with excluding some features and tests related to Mono. Some package version numbers had to be updated.

### CODE REVIEW

File: dir.targets [hunk 1]
```diff
@@ -8,11 +8,16 @@
...
+      <Error Condition="'$(MonoBuild)' == 'true' and '$(NetCoreBuild)' == 'true'"
+             Text="Cannot do both Mono and NetCore build."
+      />
+
     <Message Importance="High" Text="Build '$(MSBuildProjectFile)' into binary directory '$(BaseOutputPathWithConfig)'" />
   </Target>
 
   <Target Name="_CopyCompilers"
-          AfterTargets="_RestoreBuildToolsWrapper"
+          AfterTargets="_RestoreBuildToolsWrapper1"
```
Comment (cdmihai) [inline dir.targets]: Does `_RestoreBuildToolsWrapper1` exist? I could not find it in the repository

File: src/dir.targets [hunk 2]
```diff
@@ -9,15 +9,15 @@
 
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" Condition="('$(NetCoreBuild)' != 'true' or '$(OsEnvironment)'!='Windows_NT') and Exists('$(MSBuildToolsPath)\Microsoft.CSharp.targets')"/>
   <Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" Condition="'$(NetCoreBuild)' == 'true' and '$(OsEnvironment)'=='Windows_NT'"/>
-  <Import Condition="'$(IsTraversalBuild)'!='true' and '$(NetCoreBuild)' == 'true' and Exists('$(CompilerToolsDir)\..\build\Microsoft.Net.Compilers.props')" Project="$(CompilerToolsDir)\..\build\Microsoft.Net.Compilers.props"/>
+  <Import Condition="'$(IsTraversalBuild)'!='true' and Exists('$(CompilerToolsDir)\..\build\Microsoft.Net.Compilers.props')" Project="$(CompilerToolsDir)\..\build\Microsoft.Net.Compilers.props"/>
```
Comment (cdmihai) [inline src/dir.targets]: My pending PR #341 will remove this line (and it's equivalent in samples/dir.targets). Will this break things for !Windows?

File: dir.props [hunk 3]
```diff
@@ -149,6 +153,9 @@
   <PropertyGroup>
     <PlatformTarget>$(RuntimeArchitecture)</PlatformTarget>
     <ImportGetNuGetPackageVersions Condition="'$(OS)' != 'Windows_NT'">false</ImportGetNuGetPackageVersions>
+    <NuGetTargetMoniker Condition="'$(NetCoreBuild)' == 'true'">DNXCore,Version=v5.0</NuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(MonoBuild)' == 'true'">.NETFramework,Version=v4.5.1</NuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(NetCoreBuild)' != 'true' and '$(MonoBuild)' != 'true'">.NETFramework,Version=v4.6</NuGetTargetMoniker>
```
Comment (rainersigwald) [inline dir.props]: Really?  Aren't we targeting 4.5.1 for full-framework builds?

File: dir.props [hunk 4]
```diff
@@ -186,16 +192,16 @@
   <PropertyGroup Condition="'$(NetCoreSurface)' != 'true'">
     <DefineConstants>$(DefineConstants);FEATURE_APARTMENT_STATE</DefineConstants>
     <DefineConstants>$(DefineConstants);FEATURE_APM</DefineConstants>
-    <DefineConstants>$(DefineConstants);FEATURE_APPDOMAIN</DefineConstants>
-    <FeatureAppDomain>true</FeatureAppDomain>
+    <DefineConstants Condition="'$(MonoBuild)' != 'true'">$(DefineConstants);FEATURE_APPDOMAIN</DefineConstants>
+    <FeatureAppDomain Condition="'$(MonoBuild)' != 'true'">true</FeatureAppDomain>
```
Comment (rainersigwald) [inline dir.props]: Hmm.  I don't love this mixing--should we have a separate block for the Mono defines?

On the other hand, they should be *mostly* the same as the full-framework defines, and I don't love having to double-list everything, either . . .

File: dir.props [hunk 5]
```diff
@@ -25,7 +25,7 @@
    <!-- Common repo directories -->
   <PropertyGroup>
     <BuildToolsVersion>1.0.25-prerelease-00115</BuildToolsVersion>
-    <CompilerToolsVersion>1.0.0</CompilerToolsVersion>
+    <CompilerToolsVersion>1.1.0-rc1</CompilerToolsVersion>
```
Comment (rainersigwald) [inline dir.props]: Looks like we need this for Update 1 compatibility anyway, so--cool.

File: dir.targets [hunk 6]
```diff
@@ -81,11 +87,11 @@
          with state.  They were passing because MSTest was serial. -->
 
     <!-- Run tests on the full framework -->
-    <Exec Command="$(PackagesDir)\xunit.runner.console.$(XunitVersion)\tools\xunit.console.x86.exe @(MainAssembly, ' ') -noshadow -parallel none -xml %(MainAssembly.FullPath)_TestResults.xml -html %(MainAssembly.FullPath)_TestResults.html $(XunitOptions) > %(MainAssembly.FullPath)_stdout.txt"
+    <Exec Command="$(PackagesDir)\xunit.runner.console.$(XunitVersion)\tools\xunit.console.exe @(MainAssembly, ' ') $(XunitOptions) -noshadow -parallel none -xml %(MainAssembly.FullPath)_TestResults.xml -html %(MainAssembly.FullPath)_TestResults.html > %(MainAssembly.FullPath)_stdout.txt"
```
Comment (rainersigwald) [inline dir.targets]: I thought the x86 version was required for full-framework tests on Windows.  But clearly the tests pass.  Was that a deliberate fix on your part?

File: override.targets [hunk 7]
```diff
@@ -10,6 +10,8 @@
   <Target Name="ResolveNuGetPackageAssets" Condition="Exists('$(ProjectLockFile)') and '$(OsEnvironment)'!='Windows_NT'">
     <PropertyGroup>
       <_ExplicitReference/>
+      <_OmitTransitiveReferences Condition="'$(NetCoreBuild)'=='true'">false</_OmitTransitiveReferences>
+      <_OmitTransitiveReferences Condition="'$(NetCoreBuild)'!='true'">true</_OmitTransitiveReferences>
```
Comment (rainersigwald) [inline override.targets]: Can you comment this?

File: override.targets [hunk 8]
```diff
@@ -17,11 +19,31 @@
...
+      <!-- Remove simple name references that are already implicitly added -->
+      <_ReferencesFromNuGetPackages Remove="%(ReferencePath.FileName)" Condition="'%(ReferencePath.ResolvedFrom)' == 'ImplicitlyExpandTargetFramework'" />
+
+      <Reference Include="@(_ReferencesFromNuGetPackages)" />
+    </ItemGroup>
+    <PropertyGroup Condition=" '$(AutoUnifyAssemblyReferences)' == 'true' ">
+      <!-- Normally Design Time Assembly Resolution (DTAR) won't consider these references.
+           Put DTAR in a mode where it will prefer the output of RAR and unify. -->
+      <DTARUseReferencesFromProject>true</DTARUseReferencesFromProject>
+    </PropertyGroup>
```
Comment (rainersigwald) [inline override.targets]: Comment the origin of this logic (at least in the commit).

File: src/.nuget/packageLoad.targets [hunk 9]
```diff
@@ -40,6 +40,7 @@
     <NuGetPackageRestoreOptions Condition="'$(OsEnvironment)'!='Windows_NT'">$(NuGetPackageRestoreOptions) -PackagesDirectory "$(PackagesCache)"</NuGetPackageRestoreOptions>
     <NuGetRestoreCommand>$(NuGetToolPath) restore $(NuGetCommonOptions)</NuGetRestoreCommand>
     <NuGetInstallCommand>$(NuGetToolPath) install $(NuGetCommonOptions)</NuGetInstallCommand>
+    <NuGetInstallCommand Condition="'$(OsEnvironment)'!='Windows_NT' and '$(MonoBuild)' == 'true'">export MONO_THREADS_PER_CPU=2000; mono $(NuGetInstallCommand)</NuGetInstallCommand>
```
Comment (rainersigwald) [inline src/.nuget/packageLoad.targets]: Since it's just for one invocation you could just do:
```sh
MONO_THREADS_PER_CPU=2000 mono $(NuGetInstallCommand)
```

But it's probably not worth changing.

File: src/Shared/StrongNameHelpers.cs [hunk 10]
```diff
@@ -25,20 +26,32 @@ internal static class StrongNameHelpers
...
+
+                    if (s_GetRuntimeInterfaceAsObjectMethod != null)
+                    {
+                        s_StrongName = (IClrStrongName)s_GetRuntimeInterfaceAsObjectMethod.Invoke(null,
+                            new object[]
+                            {
+                                new Guid("B79B0ACD-F5CD-409b-B5A5-A16244610B92"),
+                                new Guid("9FD93CCF-3280-4391-B3A9-96E1CDE77C8D")
+                            });
+                    }
```
Comment (rainersigwald) [inline src/Shared/StrongNameHelpers.cs]: Why is this necessary?

File: targets/LinuxDeployDependencies.proj [hunk 11]
```diff
@@ -0,0 +1,119 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="ProcessJsonFiles" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="DeployDependencies.proj" />
```
Comment (rainersigwald) [inline targets/LinuxDeployDependencies.proj]: It's a little unusual to import a `.proj` file.  Can we avoid that?

File: targets/LinuxDeployDependencies.proj [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+          AfterTargets="RestoreRuntimePackagesDnu"
+  >
+    <Copy SourceFiles="@(IntermediateRuntimeProjectJsonLockFile)"
+          DestinationFiles="@(RuntimeProjectJsonLockFile)"
+          SkipUnchangedFiles="true"
+    />
+    
+    <!-- It appears that NuGet doesn't re-write the project.lock.json file if the contents would be the same. This can mess up incremental builds
+        if you make a non-significant change to project.json.  To avoid that, update the timestamp of the lock file here. -->
+    <Touch Files="@(RuntimeProjectJsonLockFile)" />
```
Comment (rainersigwald) [inline targets/LinuxDeployDependencies.proj]: I'd rather avoid duplicating this stuff if possible--common `.targets` imported by two projects?

File: dir.targets (hunk 1 above)
Comment (ValMenn) [inline dir.targets]: This basically disabled the targets. I am not ready to completely remove it yet, just in case... But I'll add a comment so we don't forget to do it when we are ready (the compilers are copied into the output using runtime.project.json -- we don't need to do it here.

File: src/dir.targets (hunk 2 above)
Comment (ValMenn) [inline src/dir.targets]: We'll merge. Your change should win.

File: dir.props (hunk 3 above)
Comment (ValMenn) [inline dir.props]: This is used only to restore runtime. Since there is no runtime for full framework build, the result is the same either way. Changing to 4.5.1 unfortunately causes NuGet.exe to fail (trying to get 4.5.1 framework for Ubuntu runtime)

File: dir.props (hunk 4 above)
Comment (ValMenn) [inline dir.props]: I did not think there are enough of the Mono-specific entries. Thoughts welcome.

File: dir.props (hunk 5 above)
Comment (ValMenn) [inline dir.props]: Note that this is not too useful -- only controls the target files copied. Binaries are actually copied based on the setting in targets\runtime.project.json, which is 1.1.0-beta-1, and should be OK. Updating to a higher version is problematic since it would require updates to other packages. Will do in separate PR.

File: dir.targets (hunk 6 above)
Comment (ValMenn) [inline dir.targets]: Please tell me why. It seems to work fine. I was not sure why we used x86 version so I changed it.

File: src/Shared/StrongNameHelpers.cs (hunk 10 above)
Comment (ValMenn) [inline src/Shared/StrongNameHelpers.cs]: Not available on Mono. Used only once, did not deserve an ifdef.

File: targets/LinuxDeployDependencies.proj (hunk 11 above)
Comment (ValMenn) [inline targets/LinuxDeployDependencies.proj]: Made it as a temp solution for Linux (as we have no blessed :raised_hands: Assets task there). This file can just be dropped when we get a common task from the NuGet team.

**Comment by ValMenn** [comment]
Addressed some comments

File: dir.targets (hunk 6 above)
Comment (rainersigwald) [inline dir.targets]: I have only hazy memories of failures that went away when forcing x86.  I even commented it (first line above)--please remove that.

I'd rather not use x86 if we can avoid it, so :+1:

File: dir.props (hunk 4 above)
Comment (rainersigwald) [inline dir.props]: Let's go with this for now and see how it evolves . . .

**Comment by rainersigwald** [comment]
LGTM

**Comment by cdmihai** [comment]
LGTM

File: dir.targets (hunk 6 above)
Comment (ValMenn) [inline dir.targets]: I'll revert this change. First, about 11 extra tests fail with x64 runner. Second, I think if you run under x86 you may want to use x86 runner anyway.

Comment (rainersigwald) [inline dir.targets]: Drat.  Sounds like a plan.
